client/grpc/tmservice/service.go

```diff
@@ -16,20 +18,31 @@ import (
 	"github.com/cosmos/cosmos-sdk/version"
 )
 
-// This is the struct that we will implement all the handlers on.
-type queryServer struct {
-	clientCtx         client.Context
-	interfaceRegistry codectypes.InterfaceRegistry
-}
+var (
+	_ ServiceServer                      = queryServer{}
+	_ codectypes.UnpackInterfacesMessage = &GetLatestValidatorSetResponse{}
+)
 
-var _ ServiceServer = queryServer{}
-var _ codectypes.UnpackInterfacesMessage = &GetLatestValidatorSetResponse{}
+type (
+	abciQueryFn = func(abci.RequestQuery) abci.ResponseQuery
+
+	queryServer struct {
+		clientCtx         client.Context
+		interfaceRegistry codectypes.InterfaceRegistry
+		queryFn           abciQueryFn
+	}
+)
 
 // NewQueryServer creates a new tendermint query server.
-func NewQueryServer(clientCtx client.Context, interfaceRegistry codectypes.InterfaceRegistry) ServiceServer {
+func NewQueryServer(
+	clientCtx client.Context,
+	interfaceRegistry codectypes.InterfaceRegistry,
+	queryFn abciQueryFn,
```

Comment:
still thinking about the API breaking change.

`clientCtx` has a `QueryABCI` method. Can we use that instead of the `queryFn` arg here?

Reply:
No, because that directs the query to Tendermint, which is exactly what we want to **avoid**. We want the query to go through directly via the application :)

Reply:
> which is exactly what we want to avoid

Do you know why? My understanding of #11012 is that relayers want the `/abci_query` equivalent but in gRPC. Whether the implementation routes via TM or the app should matter little.

The `tmservice` grpc service was initially thought as a gRPC proxy service over tendermint. If your PR goes directly through the app, I'm not 100% convinced this endpoint should go in `tmservice`.

Reply:
Because we're not confident Tendermint handles the queries effectively in terms of performance, locking, app degradation. I could very well be wrong on this point now.

Is this line of reasoning correct @marbar3778 @jackzampolin?

Reply:
Got it 👍. Could we add a proto comment that this endpoint endpoint does not route through tendermint?